Rework the surface upgrade tool to inform users without blocking

YuriSizov opened the pull request:
This removes the immediate confirmation dialog and insteads prints the message to the editor log (and it also appears as a toast). The immediate dialog is a devil's plaything, and it cannot be used in this scenario (if it can be used anywhere at all). The condition that triggers the SUT can happen during any attempt by the rendering server to read a mesh. This means it will conflict with a number of editor processes, like loading, importing, preview generation, export, CLI mode, etc.

So while this is less on the nose as far as informing users goes, it's also our best option to use the log and the toaster.

Closes https://github.com/godotengine/godot/issues/85202.
Closes https://github.com/godotengine/godot/issues/85201.
Likely closes https://github.com/godotengine/godot/issues/84732.

I will test a bunch of previously affected projects to see if there are no regressions. In the meantime, I'd appreciate if others double-checked what they can.

YuriSizov committed:
Rework the surface upgrade tool to inform users without blocking

This removes the immediate confirmation dialog and insteads prints the
message to the editor log (and it also appears as a toast). The immediate dialog
is a devil's plaything, and it cannot be used in this scenario (if it can be used
anywhere at all). The condition that triggers the SUT can happen during any
attempt by the rendering server to read a mesh. This means it will conflict
with a number of editor processes, like loading, importing, preview
generation, export, CLI mode, etc.

So while this is less on the nose as far as informing users goes, it's also
our best option to use the log and the toaster.

akien-mga commented:
Tested on the GDQuest TPS demo, it seems to work well. Testing procedure:
- Reset repo
- Import in 4.1.3
- Open in 4.2-rc with this PR merged

There's a slight UX issue with the toaster, as it only shows a small part of the message which doesn't give enough info. The Output panel is not expanded by default, so it's not obvious what the user should do to understand that message:

![image](https://github.com/godotengine/godot/assets/4701338/6bde89f6-4475-427f-a1b9-55cd9c056ef8)

We don't really have time to rework the toaster itself, but maybe we could push a different message to the toaster (shorter, referring to Output dock) and to the Output dock (all details)?

YuriSizov commented:
> We don't really have time to rework the toaster itself, but maybe we could push a different message to the toaster (shorter, referring to Output dock) and to the Output dock (all details)?

I think we can, I'll take a look.

Calinou commented:
Alternatively, can the toast message be wrapped over several lines?

Calinou commented on editor/surface_upgrade_tool.cpp:
```diff
@@ -170,7 +170,17 @@ SurfaceUpgradeTool::SurfaceUpgradeTool() {
+}
+
+void SurfaceUpgradeDialog::popup_on_demand() {
+	const String confirmation_message = TTR("The mesh format has changed in Godot 4.2, which affects both imported meshes and meshes authored inside of Godot. The engine needs to update the format in order to use those meshes.\n\nIf your project predates Godot 4.2 and contains meshes we recommend you run this one time conversion tool. This update will restart the editor and may take several minutes. ...
```
```suggestion
	const String confirmation_message = TTR("The mesh format has changed in Godot 4.2, which affects both imported meshes and meshes authored inside of Godot. The engine needs to update the format in order to use those meshes.\n\nIf your project predates Godot 4.2 and contains meshes, we recommend you run this one time conversion tool. This update will restart the editor and may take several minutes. Upgrading will make the meshes incompatible with previous versions of Godot.\n\nYou can still use your existing meshes as is. The engine will update each mesh in memory, but the update will not be saved. Choosing this option will lead to slower load times every time this project is loaded.");
```

Calinou reviewed (approved):
Tested locally, it works as expected on:

- https://github.com/godotengine/tps-demo
- https://github.com/godotengine/godot-demo-projects/tree/master/3d/global_illumination
- https://github.com/godotengine/godot-demo-projects/tree/master/3d/truck_town
- https://github.com/godotengine/godot-demo-projects/tree/master/3d/platformer
- https://github.com/Calinou/godot-reflection

All these projects were imported with 4.1.3 beforehand.

I notice unknown UID errors during the upgrade process though:

```
WARNING: res://stage/meshes/trunk_floor.res: In external resource #0, invalid UID: uid://7qqlq0kheobb - using text path instead: res://stage/tile_material.tres
     at: open (./core/io/resource_format_binary.cpp:1056)
WARNING: res://stage/meshes/trunk.res: In external resource #0, invalid UID: uid://7qqlq0kheobb - using text path instead: res://stage/tile_material.tres
     at: open (./core/io/resource_format_binary.cpp:1056)
WARNING: res://stage/meshes/trunk_deco.res: In external resource #0, invalid UID: uid://7qqlq0kheobb - using text path instead: res://stage/tile_material.tres
     at: open (./core/io/resource_format_binary.cpp:1056)
WARNING: res://stage/meshes/tree_top.res: In external resource #0, invalid UID: uid://7qqlq0kheobb - using text path instead: res://stage/tile_material.tres
     at: open (./core/io/resource_format_binary.cpp:1056)
WARNING: res://stage/meshes/ceiling_corner.res: In external resource #0, invalid UID: uid://7qqlq0kheobb - using text path instead: res://stage/tile_material.tres
     at: open (./core/io/resource_format_binary.cpp:1056)
WARNING: res://stage/meshes/ceiling.res: In external resource #0, invalid UID: uid://7qqlq0kheobb - using text path instead: res://stage/tile_material.tres
     at: open (./core/io/resource_format_binary.cpp:1056)
WARNING: res://stage/meshes/corner.res: In external resource #0, invalid UID: uid://7qqlq0kheobb - using text path instead: res://stage/tile_material.tres
     at: open (./core/io/resource_format_binary.cpp:1056)
WARNING: res://player/player.glb: In external resource #0, invalid UID: uid://cmvnecqga07e5 - using text path instead: res://player/player_gray.tres
     at: open (./core/io/resource_format_binary.cpp:1056)
WARNING: res://player/player.glb: In external resource #1, invalid UID: uid://be3ssftlblkjn - using text path instead: res://player/player_glow.tres
     at: open (./core/io/resource_format_binary.cpp:1056)
```

These don't appear after closing and reopening the upgraded project.

YuriSizov commented:
> Alternatively, can the toast message be wrapped over several lines (and be trimmed much later for _really_ long messages)?

Toasts respect hardcoded newlines, but don't do wrapping. I wouldn't want to get into that right now. IIRC there were cascading sizing issues in the toasts specifically.

----

I've split the message into one that goes into the log and one that's displayed as a toast. Hope it's better.

I also removed suppression when exporting because we don't need it anymore, and it was actually suppressing a bit too hard. Basically, warnings will appear as often as before if you clicked the "Update only" button.

I checked Kenney's starter project (for migration issues) and the XR template (for the export scenario). They both appear to work as expected.

YuriSizov commented:
I believe this happens already if you cancel the immediate confirmation dialog (the only difference is that we may suppress it on the initial load, but if you close and reopen it — it should still display the same amount of warnings as you currently see in this PR).

akien-mga reviewed (approved):
Tested, works as intended.

akien-mga commented:
Thanks! Amazing work :tada:

Chubercik commented:
What a great branch name 😅
